### ISSUE TITLE
inlineCriticalCss not working with SSR. 

### ISSUE THREAD

**Issue by ericfowler303** [angular/universal#2112]
# 🐞 Bug report

#### What modules are related to this issue?

- [ ] aspnetcore-engine
- [ ] builders
- [X] common
- [X] express-engine
- [ ] hapi-engine

### Is this a regression?
not sure

### Description

With inlineCriticalCss set to true, the below error occurs. I found that in 'embedLinkedStylesheet' the href value is undefined when it tries to process this element;
`<link rel="stylesheet" href="styles.162799855ab4c03c113b.css">`

Since the file path isn't a url the 'href' value is undefined. Not sure if this is a change that needs to happen in Critters or in Angular. I did find this similar issue in a different project: https://github.com/GoogleChromeLabs/critters/issues/71

## 🔥 Exception or Error

<pre><code>
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined
    at validateString (internal/validators.js:124:11)
    at Object.resolve (path.js:139:9)
    at CrittersExtended.getCssAsset (C:\Users\X\repos\X\dist\Project\server\main.js:137185:31)
    at CrittersExtended.embedLinkedStylesheet (C:\Users\X\repos\X\dist\Project\server\main.js:137265:37)
    at C:\Users\X\repos\X\dist\Project\server\main.js:137114:79
    at Array.map (<anonymous>)
    at C:\Users\X\repos\X\dist\Project\server\main.js:137114:61
    at CrittersExtended.process (C:\Users\X\repos\X\dist\Project\server\main.js:137116:8)
    at InlineCriticalCssProcessor.<anonymous> (C:\Users\X\repos\X\dist\Project\server\main.js:137609:44)
    at Generator.next (<anonymous>)

</code></pre>

## 🌍 Your Environment

<pre><code>

     _                      _                 ____ _     ___ 
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | | 
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | | 
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 12.0.0
Node: 14.15.1
Package Manager: npm 6.14.8
OS: win32 x64

Angular: 12.0.0
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, localize, material, platform-browser
... platform-browser-dynamic, platform-server, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1200.0
@angular-devkit/build-angular   12.0.0
@angular-devkit/core            12.0.0
@angular-devkit/schematics      12.0.0
@nguniversal/builders           12.0.0
@nguniversal/common             12.0.0
@nguniversal/express-engine     12.0.0
@schematics/angular             12.0.0
rxjs                            6.6.2
typescript                      4.2.4

</code></pre>

**Comment by alan-agius4** [comment]
This seems like a bug but we'll need to look at a reproduction to find and fix the problem. Can you setup a minimal repro please?

You can read [here](https://github.com/angular/universalblob/master/CONTRIBUTING.md#-submitting-an-issue) why this is needed. A good way to make a minimal repro is to create a new app via `ng new repro-app` and adding the minimum possible code to show the problem. Then you can push this repository to github and link it here.

This might be related to your directory structure so its really important to get an accurate repro to diagnose this.

**Comment by ericfowler303** [comment]
Still working on a minimal repro. I'm trying to start with a fresh angular project, add-in universal, then move over more and more which I think will have the issue happen. So far I haven't been able to reproduce in a fresh project.

There is something that I notice as different from a new project with styles vs my project which I believe started with Angular v7 and has been upgraded through versions.

In the HEAD tag, the way the main styles.css file is included is different. Maybe just resolving this could prevent the issue. Now in my problem project it's included as:
`<link rel="stylesheet" href="styles.css">`
Vs. in the fresh project it's included as:
`<link rel="stylesheet" href="styles.css" media="all" onload="this.media='all'">`
`<noscript><link rel="stylesheet" href="styles.css"></noscript>`

I did a comparison between both angular.json files and ensured that in the problem project that "optimization": true which is the same as the brand new project. Is there a different place where I should check as to why that step is happening differently?

**Comment by ericfowler303** [comment]
@alan-agius4 I believe I have figured out the issue. I have a custom caching layer in the rendering process, mostly to save cached outputs of already rendered pages for a specified time.

I realized that I wasn't passing through every renderOption through as I should. The main one I forgot here was 'publicPath' which caused different behavior here potentially:
https://github.com/angular/universal/blob/master/modules/common/engine/src/engine.ts#L68

After passing through the 'publicPath' RenderOption, I was able to get the proper async CSS and no errors when rending with 'inlineCriticalCss' set to true. Looks as this was an issue in my custom implantation, not an issue with the new inlineCriticalCss functionality.
